### PR TITLE
[Flashcard] View Flashcard

### LINKED ISSUE #137: [1.x] View a card
### Description

A user can view a card.

**Acceptance Criteria**

- [x] A user can view a card

### Example

_No response_

### PR DESCRIPTION
**Pull request by mgamal92**
| Q             | A                                                                  |
|---------------|--------------------------------------------------------------------|
| Branch?       | 1.x                                              |
| Bug fix?      |no                                                             |
| New feature?  | yes                   |
| Deprecations? | no  |
| Tickets       | Fix #137            |
| License       | MIT                                                                |

### CODE REVIEW

File: src/Controller/FlashcardController.php [hunk 1]
```diff
@@ -6,11 +6,19 @@
...
+use App\Siklid\Document\Flashcard;
 use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\Routing\Annotation\Route;
 
 class FlashcardController extends ApiController
 {
+    #[Route('/flashcards/{id}', name: 'flashcard_show', methods: ['GET'])]
+    public function show(ViewFlashcard $action, Flashcard $flashcard): JsonResponse
+    {
+        return $this->ok($action->setFlashcard($flashcard)->execute(), ['box:show', 'resource:read']);
```
Comment (imdhemy) [inline src/Controller/FlashcardController.php:19]: The `Box` document has no `box:show` serialization group. It should be `box:read` instead.

File: tests/Feature/Flashcard/CreateFlashcardTest.php [hunk 2]
```diff
@@ -6,13 +6,15 @@
...
 use App\Tests\Concern\Factory\BoxFactoryTrait;
+use App\Tests\Concern\Factory\FlashcardFactoryTrait;
 use App\Tests\Concern\WebTestCaseTrait;
 use App\Tests\TestCase;
 
 class CreateFlashcardTest extends TestCase
 {
     use WebTestCaseTrait;
     use BoxFactoryTrait;
+    use FlashcardFactoryTrait;
```
Comment (imdhemy) [inline tests/Feature/Flashcard/CreateFlashcardTest.php:17]: This trait is not used anywhere in this class.

File: tests/Feature/Flashcard/ViewFlashcardTest.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+use App\Tests\Concern\Factory\FlashcardFactoryTrait;
+use App\Tests\Concern\WebTestCaseTrait;
+use App\Tests\TestCase;
+
+class ViewFlashcardTest extends TestCase
+{
+    use WebTestCaseTrait;
+    use FlashcardFactoryTrait;
+
+    public function user_can_show_single_flashcard(): void
```
Comment (imdhemy) [inline tests/Feature/Flashcard/ViewFlashcardTest.php:17]: This test does not run. A `@test` comment is required.

File: tests/Feature/Flashcard/ViewFlashcardTest.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        $client = $this->makeClient();
+        $user = $this->makeUser();
+        $flashcard = $this->makeFlashcard(['front' => 'front', 'back' => 'back', 'user' => $user]);
+        $client->loginUser($user);
+
+        $client->request('GET', '/api/v1/flashcard/'.$flashcard->getId());
+
+        $this->assertExists(Flashcard::class, ['id' => $flashcard->getId()]);
```
Comment (imdhemy) [inline tests/Feature/Flashcard/ViewFlashcardTest.php:26]: Assertion of existence has nothing to add here. We need to assert the returned response contains the requested flashcard.

File: src/Controller/FlashcardController.php [hunk 5]
```diff
@@ -6,11 +6,19 @@
...
 use App\Siklid\Application\Flashcard\CreateFlashcard;
+use App\Siklid\Application\Flashcard\ViewFlashcard;
+use App\Siklid\Document\Flashcard;
 use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\Routing\Annotation\Route;
 
 class FlashcardController extends ApiController
 {
+    #[Route('/flashcards/{id}', name: 'flashcard_show', methods: ['GET'])]
+    public function show(ViewFlashcard $action, Flashcard $flashcard): JsonResponse
```
Comment (imdhemy) [inline src/Controller/FlashcardController.php:17]: Parameter conversion looks cleaner somehow. But, I'm still not sure, if should we expose Documents to the Controllers or just expose actions as a single entry point to the business logic.

**Review by imdhemy — CHANGES_REQUESTED** [review]
Thanks a lot for investing time in this PR. I added some comments. Feel free to discuss about the.

File: tests/Feature/Flashcard/CreateFlashcardTest.php (hunk 2 above)
Comment (mgamal92) [reply tests/Feature/Flashcard/CreateFlashcardTest.php]: Done

File: tests/Feature/Flashcard/ViewFlashcardTest.php (hunk 3 above)
Comment (mgamal92) [reply tests/Feature/Flashcard/ViewFlashcardTest.php:20]: Done

File: src/Controller/FlashcardController.php (hunk 5 above)
Comment (mgamal92) [reply src/Controller/FlashcardController.php:17]: I didn't get that point

File: src/Controller/FlashcardController.php (hunk 1 above)
Comment (mgamal92) [reply src/Controller/FlashcardController.php:19]: Done

File: src/Controller/FlashcardController.php (hunk 5 above)
Comment (imdhemy) [reply src/Controller/FlashcardController.php:17]: This is a tradeoff anyway. The current approach is fine. Let's discuss this after fixing the feature test.

**Comment by mgamal92** [comment]
@imdhemy check it

**Review by imdhemy — APPROVED** [review]
@mgamal92 Thank you!

**Pull request merged**